### PR TITLE
Bit 82 grpc security

### PR DESCRIPTION
**Pull request by Eugene-hu**
Major Changes:
-Axon server security: messages now must contain a signature and a nounce for requests to be accepted
-Receptor signing: Receptor messages are now signed using the encrypted hotkey of the wallet
-axon unit tests updated to reflect new functionality

Minor Changes:
-qqdm fix
-added pubkey property for wallet
-bittensor version added as part of the metadata

### CODE REVIEW

File: bittensor/_axon/__init__.py [hunk 1]
```diff
@@ -180,4 +185,62 @@ def check_forward_callback( forward_callback:Callable, modality:int, pubkey:str
...
+                data_time = datetime.strptime(nounce,'%m%d%Y%H%M%S%f')
+                _keypair = self.keypair(ss58_address=pubkey)
+                
+                #checking the time of creation, compared to previous messages
+                if pubkey in self.nounce_dic.keys():
+                    prev_data_time = self.nounce_dic[pubkey]
+                else:
+                    self.nounce_dic[pubkey] = data_time
+                    verification = _keypair.verify(nounce+pubkey,message)
+                    if verification:
```
Comment (unconst) [inline bittensor/_axon/__init__.py]: what happens is this does not pass?

File: bittensor/_axon/__init__.py [hunk 2]
```diff
@@ -180,4 +185,62 @@ def check_forward_callback( forward_callback:Callable, modality:int, pubkey:str
...
+                        return continuation(handler_call_details)
+
+                if data_time - prev_data_time >= timedelta(milliseconds=1):
```
Comment (unconst) [inline bittensor/_axon/__init__.py]: I think you need prev_data_time but this does not exist unless the first if statement passes

File: bittensor/_axon/__init__.py [hunk 3]
```diff
@@ -180,4 +185,62 @@ def check_forward_callback( forward_callback:Callable, modality:int, pubkey:str
...
+        self._valid_metadata = ('rpc-auth-header', key)
+        self.keypair = keypair
+        self.nounce_dic = {}
+
+        def deny(_, context):
+            context.abort(grpc.StatusCode.UNAUTHENTICATED, 'Invalid key')
+
+        self._deny = grpc.unary_unary_rpc_method_handler(deny)
+
+    def intercept_service(self, continuation, handler_call_details):
```
Comment (unconst) [inline bittensor/_axon/__init__.py:208]: Can we write this function with less nesting, I see 5 levels here which makes it hard to debug and understand

**Review by unconst — COMMENTED** [review]
Minor fixes. Overall awesome work.

**Review by shibshib — APPROVED** [review]

File: bittensor/_axon/__init__.py [hunk 4]
```diff
@@ -83,7 +85,10 @@ def __new__(
         if thread_pool == None:
             thread_pool = futures.ThreadPoolExecutor( max_workers = config.axon.max_workers )
         if server == None:
-            server = grpc.server( thread_pool, maximum_concurrent_rpcs = config.axon.maximum_concurrent_rpcs )
+            server = grpc.server( thread_pool,
+                                  interceptors=(AuthInterceptor(Keypair),),
```
[the remark this answers is not in the document]
Comment (Eugene-hu) [reply bittensor/_axon/__init__.py]: Yes, a bit of legacy code here; gonna rewrite it so it is clearer for the reader.

File: bittensor/_axon/__init__.py [hunk 5]
```diff
@@ -180,4 +185,62 @@ def check_forward_callback( forward_callback:Callable, modality:int, pubkey:str
...
+            context.abort(grpc.StatusCode.UNAUTHENTICATED, 'Invalid key')
+
+        self._deny = grpc.unary_unary_rpc_method_handler(deny)
+
+    def intercept_service(self, continuation, handler_call_details):
+        r""" Authentication between bittensor nodes. A signature is created using
+            publickey and encodes the time at which the signature was created.
+        """
+        meta = handler_call_details.invocation_metadata
+        if meta[0] == self._valid_metadata:
```
[the remark this answers is not in the document]
Comment (Eugene-hu) [reply bittensor/_axon/__init__.py]: yeap, there is a header and a bittensor version in the metadata for us to use in the future.

File: bittensor/_axon/__init__.py [hunk 6]
```diff
@@ -180,4 +185,62 @@ def check_forward_callback( forward_callback:Callable, modality:int, pubkey:str
...
+            try: 
+                #version checking
+                if bittensor.__version_as_int__ != int(meta[2].value):
+                    self._deny
+
+                nounce, pubkey, message = meta[1].value.split('bitxx')
```
[the remark this answers is not in the document]
Comment (Eugene-hu) [reply bittensor/_axon/__init__.py]: no particular reason, needed a long string to separate the messages

File: bittensor/_wallet/wallet_impl.py [hunk 7]
```diff
@@ -209,6 +209,20 @@ def hotkeyfile(self) -> str:
         )
         return os.path.join(full_path, "hotkeys", self._hotkey_string)
 
+    @property
```
[the remark this answers is not in the document]
Comment (Eugene-hu) [reply bittensor/_wallet/wallet_impl.py]: Sounds good

File: bittensor/_axon/__init__.py (hunk 3 above)
Comment (Eugene-hu) [reply bittensor/_axon/__init__.py:208]: yea, gonna rewrite parts of the logic, so it is easier to read and understand.

**Pull request merged**